### PR TITLE
Initial code commit with basic functionality, some tests, and a README

### PR DESCRIPTION
**Pull request by srawlins**
This is the basic functionality of expected_output. It's the same functionality in use by dart_style and markdown (and others?).

### CODE REVIEW

**Comment by srawlins** [comment]
Rather than a rubber stamp approval, I'd love an actual review of, specifically, a few pieces here:

* the README. It's kind of slap dash stream of conciousness right now.
* naming
  * specifically, what are these data test case input/expectedOutput files? data files? test case files?
  * what about each "case", i.e. the input/expectedOutput pair? I have a class named `DataCase` but I just totally made this up, and I'm not married to it.
  * The API names: `dataCases`, `dataCasesUnder`. Meh?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+### Code reviews
+
+All submissions, including submissions by project members, require review.
+
+### File headers
+
+All files in the project must start with the following header.
+
+    // Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
```
Comment (pq) [inline CONTRIBUTING.md]: 2017?

File: lib/expected_output.dart [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+}) sync* {
+  var entries =
+      new Directory(directory).listSync(recursive: false, followLinks: false);
+  for (var entry in entries) {
+    if (!entry.path.endsWith(extension)) {
+      continue;
+    }
+
+    var file = p.basename(entry.path)
+    .replaceFirst(new RegExp('\.$extension\$'), '');
```
Comment (pq) [inline lib/expected_output.dart]: Nit: this formatting looks odd.  I'd expect something more like:

```
  var file =
      p.basename(entry.path).replaceFirst(new RegExp('\.$extension\$'), '');
```

?

Comment (pq) [reply lib/expected_output.dart]: Or are you shying away from `dartfmt`?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+THIS IS ANOTHER PARAGRAPH.
+```
+
+We can quickly create tests over these data cases with some Dart:
+
+```dart
+library touppercase.tests;
+
+void main() {
+  for (var dataCase in dataCasesUnder(library: #touppercase.tests)) {
```
Comment (pq) [inline README.md:59]: What's the benefit in using reflection vs. specifying a data directory?

**Review by pq — APPROVED** [review]
In general looks great.  I'm squeamish about using `mirrors` but maybe I'm just being superstitious.

**Comment by pq** [comment]
> the README. It's kind of slap dash stream of conciousness right now.

It's better than you think! 👍 

> naming
> specifically, what are these data test case input/expectedOutput files? data files? test case files?

I take this to be an implementation detail so wouldn't sweat it too much.

> what about each "case", i.e. the input/expectedOutput pair? I have a class named DataCase but I > just totally made this up, and I'm not married to it.

Same.  @munificent may have some more thoughts here since it's been a while since I wrote any tests this way ---  though now that I see it again, you've got me thinking! 😉 

> The API names: dataCases, dataCasesUnder. Meh?

Hmmmm.  As I mentioned, I'm curious about the decision to reflect rather than just specify directories.  If `mirrors` is the way to go though I think `dataCases` ain't bad.

File: lib/expected_output.dart (hunk 2 above)
Comment (srawlins) [reply lib/expected_output.dart]: Durp. Sorry I forgot to dartfmt :stuck_out_tongue_closed_eyes:

File: CONTRIBUTING.md (hunk 1 above)
Comment (srawlins) [reply CONTRIBUTING.md]: Done!

File: README.md (hunk 3 above)
Comment (srawlins) [reply README.md:59]: There is no other way to get a directory relative to the executing code. (Unless there recently is.) The mirrors incantation comes from https://github.com/dart-lang/test/issues/110.

Comment (srawlins) [reply README.md:59]: I just put a little comment in the README about this.

**Comment by srawlins** [comment]
@munificent, not urgent, but I'll wait a day or two in case you want to put in some input on naming. I'd rather fix it sooner than later if you hate the names or the API.

**Review by munificent — APPROVED** [review]
Looks nice! 👍

**Comment by srawlins** [comment]
Thanks @pq and @munificent !

**Pull request merged**